linode/manager
# refactor: [M3-7382] - Use `object-storage/types` endpoint for pricing data

## Description 📝
API is now returning pricing data through the `/object-storage/types` endpoint. 

In this PR, prices can be dynamically displayed for base region pricing and DC-specific region pricing, using the prices returned in the new endpoint, for Object Storage.

> [!NOTE]  
> Outbound transfer overages will be done in a follow-up PR, since pricing for network transfer overages is the same across Cloud Manager and not currently being returned by the API. They are creating a new endpoint for this.

## Changes  🔄
- Created the necessary api-v4 endpoint, query, and interface to use the object-storage/types endpoint.
- Modified `getDCSpecificPriceByType` to optionally accept interval (hourly or monthly) and decimal precision params and updated util unit test accordingly.
- Created the necessary factories and endpoint in the MSW to support a mocked response to the API endpoint.
- Added UX for error handling in the unlikely event that pricing cannot be calculated: disable primary action button and display tooltip error. 
- Added loading state for prices.
- Updated unit tests.

## Target release date 🗓️
5/28/24

## Preview 📷

| Before  | After   |
| ------- | ------- |
| ![Screenshot 2024-05-17 at 12 50 05 AM](https://github.com/linode/manager/assets/114685994/6bdbf681-4673-435e-b0ec-42ea8054d2b9) ![Screenshot 2024-05-17 at 12 50 23 AM](https://github.com/linode/manager/assets/114685994/4e431b16-ee31-4aeb-a947-0789bbb15cb6) ![Screenshot 2024-05-17 at 12 50 33 AM](https://github.com/linode/manager/assets/114685994/c0e5263c-9022-4843-88c5-9fa8e9594c53) | ![Screenshot 2024-05-20 at 8 08 52 AM](https://github.com/linode/manager/assets/114685994/0938106c-b6af-47b3-bc3f-36f1feaf7cc7) ![Screenshot 2024-05-17 at 12 47 38 AM](https://github.com/linode/manager/assets/114685994/0eb5f0fd-a8d4-4d35-8ba3-059d4603f37d) ![Screenshot 2024-05-17 at 12 47 49 AM](https://github.com/linode/manager/assets/114685994/b445d5a8-868d-4e1f-a0de-e4eb10c4016a) |
| N/A | <video src="https://github.com/linode/manager/assets/114685994/c2551927-a383-4fda-a8b7-afa4cc026501"/>|
| ![Screenshot 2024-05-17 at 12 50 46 AM](https://github.com/linode/manager/assets/114685994/d9fa5437-dfd1-4977-bd86-206588c3a48f) | ![Screenshot 2024-05-17 at 12 46 48 AM](https://github.com/linode/manager/assets/114685994/77f6fd78-ae68-43dd-86c2-b3df6b862afc) |
| ![Screenshot 2024-05-17 at 12 51 03 AM](https://github.com/linode/manager/assets/114685994/ed1b47a6-7d22-4ce3-928d-f2750d1f831d) | ![Screenshot 2024-05-17 at 12 47 10 AM](https://github.com/linode/manager/assets/114685994/8b7172c1-a951-4af6-8ddb-71c39c12dce8) |


## How to test 🧪

### Prerequisites
(How to setup test environment)
- Check out this PR and `yarn up`.
- Go to http://localhost:3000/account/settings and Cancel Object Storage if it is enabled on your account.

### Verification steps
(How to verify changes)
- Go to http://localhost:3000/object-storage/buckets/create.
- Select a region without DC specific pricing (e.g. `us-east`).
- Open the network tab and filter on `/types`. Confirm the "additional storage overage" pricing is consistent with the overage pricing returned by the API endpoint AND that price is consistent with prod.
- Repeat the same process for Sao Paulo (`br-gru`) and Jakarta (`id-cgk`).
- Fill out the necessary fields, click Create Bucket, and confirm the Enable Object Storage modal pops up with no regressions. *Cancel out of the modal - don't enable OBJ.*
- Use the Network tools to slow and block network requests - confirm that loading and error behavior is handled gracefully in the Create Bucket drawer if prices can't be retrieved from the `object-storage/types` endpoint.
- Go to http://localhost:3000/object-storage/access-keys.
- Fill out the necessary fields, click Create Access Key, and confirm the Enable Object Storage modal pops up with no regressions. *Cancel out of the modal - don't enable OBJ.*
- Ensure object storage e2e tests pass.
- Ensure updated unit tests pass:
```
yarn test OveragePricing dynamicPricing EnableObjectStorageModal
```

## As an Author I have considered 🤔

*Check all that apply*
